esp8266/modpybhspi: Add a HSPI module for hardware SPI support (MetalPhreak version)

deshipu opened the pull request:
This module uses ESP8266's SPI hardware, which allows much higher
speeds. It uses a library from https://github.com/MetalPhreak/ESP8266_SPI_Driver

deshipu committed:
esp8266/modpybhspi: Add a HSPI module for hardware SPI support

This module uses ESP8266's SPI hardware, which allows much higher
speeds. It uses a library from
https://github.com/MetalPhreak/ESP8266_SPI_Driver

deshipu commented:
I took the liberty of cleaning up MetalPhreak's code in terms of style a little: removing the gaudy comment adornments, fixing type names, correcting spacing, etc. -- I think it reads better now.

deshipu commented:
I also disabled the ``CS`` pin, because on Micropython that is always going to be handled by the driver code.

deshipu commented:
Fixed problems in how the read methods were registered, and implemented all the missing duplex functionality.

deshipu commented:
I optimized the loops in all of the functions, so that you don't have delays between bytes -- they are sent as fast as the hardware allows.

deshipu commented:
In the end I just figured out how to read the pixel data from the ili9341 display, and tested reading on that. Works fine up to 20Mhz, at higher speeds the display gives up (it's rated at 10Mhz, so still pretty good). I have no device I could read from at 80Mhz, though the ili9341 seems to accept that speed for writing just fine.

I'm now trying to get someone else to test this too.

pfalcon commented on docs/esp8266/quickref.rst:
```diff
@@ -183,6 +183,19 @@ The SPI driver is implemented in software and works on all pins::
+------------
+
+The hardware SPI is faster (up to 80Mhz), but only works on following pins:
+``MISO``=12, ``MOSI``=13, ``SCK``=14. It has the same methods as SPI, except
```
It's unclear what "12", etc. mean here. Physical chip pin number? GPIO number? (NodeMCU peoplew will think it's nodemcu logical number, etc.) It should be "GPIO12", etc. to disambiguate.

deshipu replied:
Agreed, I will fix that.

pfalcon commented on esp8266/hspi.c:
```diff
@@ -0,0 +1,353 @@
+    if (spi_no > 1) {
+        return;
+    }
+    if ((prediv==0)|(cntdiv==0)) {
```
This is not correct and not acceptable. It should be `||`.

deshipu replied:
I really wouldn't know, I didn't write this code, this is a library that I'm using, and I tried to really not change anything in it.

pfalcon replied:
But you said that you already applied some codestyle changes. Of those, this would be first to apply. I'm surprised that you didn't get any warnings on this (which are fatal in uPy), that may change any time (newer gcc version, clang, etc.)

deshipu replied:
I changed style where I was sure it didn't affect what the code does.  There is a significant difference between ``|`` and ``||``, and I don't think it's just a matter of style.

pfalcon commented on esp8266/hspi.c:
```diff
@@ -0,0 +1,353 @@
+Just do minimal work needed to send 8 bits.
+*/
+inline void spi_tx8fast(uint8_t spi_no, uint8_t dout_data) {
+    if (spi_no > 1) {
```
This is Nth repeating check. In MicroPython code, we don't need these. (Don't do anything about them so far though.)

deshipu replied:
Again, this is a library code. I suppose this check is there, because this is one of the exposed API functions.

pfalcon commented on esp8266/hspi.h:
```diff
@@ -0,0 +1,79 @@
+
+// Define SPI hardware modules
+#define SPI 0
+#define HSPI 1
```
So, what "HSPI" is, after all?

deshipu replied:
HSPI, also known as "user SPI", is the second SPI interface on the ESP8266, the one that is not taken by the flash memory, and thus is usable for the general user.

pfalcon replied:
Yeah, so what does "H" mean?

deshipu replied:
See ttp://d.av.id.au/blog/esp8266-hardware-spi-hspi-general-info-and-pinout/ for explanation. This the name they use in the documentation (like here https://www.scribd.com/doc/305609158/8q-Esp8266-Hspi-Host-Multi-Device-API-en-v1-0) I have no idea what it means, perhaps something in Chinese.

pfalcon commented on esp8266/modpybhspi.c:
```diff
@@ -0,0 +1,258 @@
+                "impossible baudrate"));
+        }
+        self->baudrate = 80000000L / (prediv * cntdiv);
+        spi_init_gpio(HSPI, SPI_CLK_USE_DIV);
```
Here and elsewhere you arbitrarily lock out one of 2 SPI hardware blocks. Doesn't make sense to do that, at least not like that. Class constructor should accept `id` argument, you should validate it once to be 0 or 1, store it in object, and pass to spi_* funcs, which then don't need to check that it's 0 or 1 again and again. *Then* you may want to add a separate check to a class that only "1" is allowed and raise exception otherwise (for other people to comment out that check later).

deshipu replied:
I don't lock it out arbitrarily. That SPI interface is used by the ESP8266 modules for communication with their on-module flash memory chip. There is no sense exposing this in Micropython, because any attempt to use them while the microcontroller is running from the flash chip, except for very narrow and carefully crafted special cases, would lead to a crash and/or corruption of data. I strongly believe that things that don't work should not be exposed, and we shouldn't give users one more way to shoot themselves in the foot.

pfalcon replied:
Per MicroPython HW API, all hardware classes take first "id" parameter, regardless whether there's only hw block or not (for consistency across ports).

deshipu replied:
I certainly didn't notice that in the ESP8266's APIs.

pfalcon commented:
Excellent work on hardware side of things! The issues as usual come on API design side. We should not pretend that there's only one hardware SPI block - there're two. And our hardware API requires for block ID to be passed in (even it's just one block, which is not the case for esp8266 anyway). I'd suggest to go and fixing that issue right away.

Further API design issue is how bitbanging SPI and hardware SPI play with each other (from API design point of view again, i.e. how user select which one to use, and how consistent is such usage). I opened https://github.com/micropython/micropython/issues/2182 on that. I do not suggest to make any changes based on that so far, but welcome comments on that (well thought out comments, as usual).

deshipu committed:
docs/esp8266/quickref: Improve the HSPI pin descriptions

Make it clear what the numbers for the pins actually refer to.

deshipu committed:
esp8266/hspi: Remove check for spi_no and some cleanup

The checks for spi_no are not needed, because we always pass the
same value.

Some more style corrections, fix the | -> || issue.

deshipu commented:
I made the changes you suggested (for now as separate commits, so that it's easier to see them, we can squash them once it's ready to merge). I'm not really a C programmer, and I've been just blindly copy-pasting things there a little, so your feedback is valuable there.

I also don't know the Micropython's C API -- the HSPI module looks the way it does, because I copied practically all method definitions from the SPI module. I wouldn't even know where to start to redefine those things to meet your specification -- is there *any* documentation on that available? I would love to learn, but I couldn't find anything.

Finally, here's a snippet from the PEP-0008 document, which I try to follow when writing code in Python:

> Some other good reasons to ignore a particular guideline:
> * When applying the guideline would make the code less readable, even for someone who is used to reading code that follows this PEP.
> * To be consistent with surrounding code that also breaks it (maybe for historic reasons) -- although this is also an opportunity to clean up someone else's mess (in true XP style).
> * Because the code in question predates the introduction of the guideline and there is no other reason to be modifying that code.
> * When the code needs to remain compatible with older versions of Python that don't support the feature recommended by the style guide.

pfalcon commented:
> Because the code in question predates the introduction of the guideline and there is no other reason to be modifying that code.

Yes, changing upstream code as little as possible is a good idea. The main reason for that would be simplify syncing with upstream in the future, 2nd is to avoid extra work. And if changing upstream code is really needed, the best way to do that is to commit it initially as is, and then apply changes in separate commits, so it was (and will be) clearly visible what changes were applied. (I didn't tell you that, as I don't really want to put more burden than needed). 

I suggested to change `|` -> `||` in this patch (and not later for example) because you already said:

> I took the liberty of cleaning up MetalPhreak's code in terms of style a little: removing the gaudy comment adornments, fixing type names, correcting spacing, etc.

pfalcon commented:
> I also don't know the Micropython's C API -- the HSPI module looks the way it does, because I copied practically all method definitions from the SPI module. I wouldn't even know where to start to redefine those things to meet your specification

I'm not sure I understand exactly what you're asking about. Let's consider it from following point of view:

I think that you did a great job. 2 things don't allow me to merge your code ASAP though (with any other changes to be done afterwards): a) that `|` vs `||` issue; b) suggestion to pass first `id` argument to constructor.

You did a). What about b)? Do you agree to do it after all? If yes, do you think that you can do it, or you miss some details how to do it?

deshipu commented:
It's both.

On one hand, I strongly disagree that the flash memory interface should be exposed to Python programmers, especially as the first, "default" interface. Any attempt to use that on a board that runs Micropython (and so must have been booted from a flash chip) has to lead to a crash and possibly data corruption. That's simply wrong, inviting people to shoot themselves in the foot, we should avoid such traps. Since you can't really use that while Micropython is running, it doesn't make sense to expose this.

On the other hand, I copied all the magical method definitions and other incantations that make those methods visible in the Python module directly from the ``modpybspi.c`` file, and I have no idea how I would even start to change their signatures. I tried that initially, and the result was a method that always made the ESP8266 crash and restart when called. I also couldn't find any documentation whatsoever on how those magical macros should be used and what they actually mean, and I'm not familiar with the intricacies of C language and internals of Micropython itself to figure this out from code.

pfalcon commented:
>  I strongly disagree that the flash memory interface should be exposed to Python programmers, especially as the first, "default" interface.

I covered that in my very first comments: "Then you may want to add a separate check to a class that only "1" is allowed and raise exception otherwise"

> Since you can't really use that while Micropython is running

It's quite different, *you can't use now*. You should be the first one to want to make that possible, so you can attach SPI stuff to SPI0, and have SPI1's pins free for something else.

But as I'm repeating for 3rd or so time, the talk is about even simpler thing - Hardware API specification requires first argument to SPI class to be `id`, that's enough of a requirement.

deshipu commented:
As I said, I have no idea how to do it. I guess I will just wait for the software SPI module to be fixed, and then I will copy that.

pfalcon commented:
Ok, so if you agree that's the right way to do, I can help with that. Software SPI fixing depends on https://github.com/micropython/micropython/issues/2182, and may take some time.

deshipu commented:
In that case, wouldn't it be better to keep the SPI and HSPI consistent, and then fix them both together when we are ready?

pfalcon commented:
Well, when I see someone submitting a patch, I assume they're interested to have it merged ASAP. If it's not the case and you're happy to wait, well, it can wait. (Checking up, the oldest pull request we have is 1.5 years old; at least I know that for any good PR, I tried to work with a submitter to get it in ASAP as it's good enough, and minimum time is indeed 5 minutes).

deshipu commented:
It's nice to get feedback ASAP, and I'm very grateful for that. However, I'd rather have merged a good patch, than have it merged fast. For my own use, I can just use this, so I'm not in a hurry.

However, I wonder if hardware SPI wouldn't be also suitable for writing more reliable drivers for the addressable LEDs.

pfalcon commented:
Hardware SPI is a great contribution and great help for maintainers (who then are able on other important and useful things in the meantime). But if you're ok with it, let's indeed wait for @dpgeorge's feedback (he's travelling now, so it may take some time).

pfalcon commented:
@dpgeorge : Let's decide what to do about this, taking into account conclusions at https://github.com/micropython/micropython/issues/2182 .

deshipu committed:
esp8266/modmachinespi: Add a factory method for SoftSPI/HSPI

pfalcon commented:
Thanks, the last commit is exactly what I meant in the other ticket. So, there may be a bit more cleanups, but that's always the case with more or less big changes. Let's wait no more, I squashed these down to 2 commits and merged. Thanks for making this!

pfalcon commented:
Well, so one of the next changes I'd like to do is to remove exposure for individual SoftSPI/HSPI classes, instead making only SPI available. Not ding that doesn't give any benefits, the only outcome could be that people will write apps specific to esp8266 and less portable to other systems.

deshipu commented:
Thank you.

deshipu commented:
Well, one potentially useful thing about making SoftSPI available is that you can then subclass it.

pfalcon commented:
What's the point of subclassing in it? In general, and specifically for MicroPython, given that there's very little support for subclassing native classes anyway.

deshipu commented:
Generally the Python community always made those kinds of decisions assuming, that we can't know in advance what use cases there may be, and that the programmers are consenting adults, know what they do, and should be empowered. I've seen a lot of propositions for removing or hiding some of the more obscure features (such as the ``__foo`` name mangling), and the decision was always to leave them be.

I understand that MicroPython has its own set of guidelines and heuristics by which such decisions are made, and I think the most important one is memory use optimization. Is there any benefit, other than hiding them from view, of hiding those classes? I guess the few bytes you save on the entries in the module locals dict, and on the name strings. Anything else?

pfalcon commented:
Yes, priorities for MicroPython are different. A general problem of Python is over-dynamicity (if someone thinks Python is slow, that's why). With MicroPython, we exchange that Pandora box for being able to run a nice VHLL on an MCU at all.

As for second part, that's not exactly the right way to pose question. There was single SPI class before your patch, and there was discussion that there should be single after it. You chose to expose more classes, why?

deshipu commented:
Because it didn't cost much, and it seemed like a logical thing to do. But I don't actually have a strong opinion on the subject. I'm playing a devil's advocate right now, just to make sure we explore the options (and there is that bad troll habit I'm trying to get rid of too).

On a slightly related note, should I follow up with a patch that cleans up the file and function names? Are there plans to drop the "pyb" name completely at some point?

pfalcon commented:
Ok, so a high-level problem MicroPython have is that it's very hard to agree on APIs beyond those which we inherit from upstream. And there's always risk that we'll never agree at all, and MicroPython will be mess like many other languages (and being a mess a typical reason for a project demise). As there's an API spec, and it says that there's machine.SPI, I'm as a maintainer, need to follow it, and make sure that's the standard interface to access core functionality, and there're no unnecessary backdoors which may lead to confusion and mess. That's my motivation.

I don't think it's worth to do more renames at this time. At the very least, I'd prefer myself/@dpgeorge to look at it first well enough and see what can be (if at all) simplified.

mcauser commented on docs/esp8266/quickref.rst:
```diff
@@ -189,6 +189,20 @@ The SPI driver is implemented in software and works on all pins::
+
+    from machine import Pin, HSPI
+
+    hspi = HSPI(baudrate=800000000, polarity=0, phase=0)
```
@deshipu I tried this and received `ValueError: impossible baudrate`
800000000 -> 800,000,000 isn't that 800Mhz? One too many zeros?

deshipu replied:
Oops, you are right! I will fix it.

dpgeorge commented:
I simplified the SPI and HSPI classes to have less code duplication in 0823c1baf864f2e8fd60e49a5e42a958fd89ff43, 5863e15a23910f46c716876e171a23008993db77 and 3be8b688c04d7210872d099d9d65269bbe5c13c5.  It saved about 500 bytes of code.

deshipu commented:
That is great, thank you! Seeing those examples really lets me learn.

dpgeorge commented:
@deshipu no problem, the refactoring helped stmhal port as well.  BTW, you now may want to try and optimise the hspi_transfer function so it doesn't need the additional loop at the end (the one without ets_loop_iter).  For example, you could just do: `if (i % chunk_size == 0) { ets_loop_iter(); }` to only call it periodically.

deshipu commented:
I tried that initially, and it resulted with very slow transfers. Seems like a loop that only has the single assignment in it gets optimized by the compiler into very efficient code. That's why I did it the way it is.
